### PR TITLE
Reorganize learning module

### PR DESCRIPTION
**Pull request by SenWu**
* Remove original ``Classifier`` class, rename ``NoiseAwareModel`` to ``Classifier`` and use the same setting for both binary and multi-class classifier
* Reorganize ``learning`` module to use pytorch dataloader, include ``MultiModalDataset`` to better handle multimodal information, and simplify the code
* Unify the loss (``SoftCrossEntropyLoss``) for all settings
* Rename ``layers`` in learning module to ``modules``
* Remove ``batch_size`` input argument from ``_calc_logits``, ``marginals``, ``predict``, ``score`` in ``Classifier``
* Rename ``predictions`` to ``predict`` in ``Classifier`` and update the input arguments to have ``pos_label`` (assign positive label for binary class prediction) and ``return_probs`` (If True, return predict probablities as well)
* Update ``score`` function in ``Classifier`` to include:
  (1) For binary: precision, recall, F-beta score, accuracy, ROC-AUC score;
  (2) For categorical: accuracy;
* Remove ``LabelBalancer``

### CODE REVIEW

File: src/fonduer/learning/disc_models/lstm.py [hunk 1]
```diff
@@ -136,7 +159,7 @@ def _update_settings(self, X):
         :type X: list of (candidate, features) pairs
         """
 
-        self.logger.info("Load defalut parameters for LSTM")
+        self.logger.info("Loading defalut parameters for LSTM")
```
Comment (lukehsiao) [inline src/fonduer/learning/disc_models/lstm.py]:
```suggestion
        self.logger.info("Loading default parameters for LSTM")
```

File: CHANGELOG.rst [hunk 2]
```diff
@@ -52,6 +52,22 @@ Fixed
 
 Changed
 ^^^^^^^
+* `@senwu`_: Reorganize ``learning`` module to use pytorch dataloader, include
+  ``MultiModalDataset`` to better handle multimodal information, and simplify 
+  the code
+* `@senwu`_: Remove ``batch_size`` input argument from ``_calc_logits``, 
+  ``marginals``, ``predict``, ``score`` in ``Classifier``
```
Comment (lukehsiao) [inline CHANGELOG.rst]:
```suggestion
  ``marginals``, ``predict``, and ``score`` in ``Classifier``
```

File: src/fonduer/learning/classifier.py [hunk 3]
```diff
@@ -36,35 +325,48 @@ def save_marginals(self, session, X, training=False):
...
+    def predict(self, X, b=0.5, pos_label=1, return_probs=False):
+        """Return numpy array of class predictions for X
         based on predicted marginal probabilities.
 
         :param X: Input data.
-        :param b: Prediction threshold.
+        :param b: Decision boundary *for binary setting only*.
         :type b: float
+        :param pos_label: Positive class index *for binary setting only*. Default: 1
+        :type pos_label: int
```
Comment (lukehsiao) [inline src/fonduer/learning/classifier.py]: Add docstring for `return_probs`?

File: src/fonduer/learning/classifier.py [hunk 4]
```diff
@@ -73,8 +375,10 @@ def score(
...
         :param b: Decision boundary *for binary setting only*.
         :type b: float
+        :param pos_label: Positive class index *for binary setting only*. Default: 1
+        :type pos_label: int
         :param set_unlabeled_as_neg: Whether to map 0 labels -> -1,
-            *binary setting.*
+            *for binary setting only*
         :type set_unlabeled_as_neg: bool
         :param beta: For F-beta score; by default beta = 1 => F-1 score.
         :type beta: int
```
Comment (lukehsiao) [inline src/fonduer/learning/classifier.py]: Docstring for `batch_size` should be removed also.

File: src/fonduer/learning/disc_models/modules/loss.py [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+    def __init__(self, weight=None, reduction="mean"):
+        super().__init__()
+        if weight is None:
+            self.weight = None
+        else:
+            self.register_buffer("weight", torch.Tensor(weight))
+
+        self.reduction = reduction
+
+    def forward(self, input, target):
```
Comment (lukehsiao) [inline src/fonduer/learning/disc_models/modules/loss.py]: Add docstring for this function.

File: src/fonduer/learning/disc_models/sparse_logistic_regression.py [hunk 6]
```diff
@@ -112,7 +137,7 @@ def _update_settings(self, X):
         :type X: list of (candidate, features) pair
         """
 
-        self.logger.info("Load defalut parameters for Sparse Logistic Regression")
+        self.logger.info("Loading defalut parameters for Sparse Logistic Regression")
```
Comment (lukehsiao) [inline src/fonduer/learning/disc_models/sparse_logistic_regression.py]:
```suggestion
        self.logger.info("Loading default parameters for Sparse Logistic Regression")
```

File: src/fonduer/learning/disc_models/sparse_lstm.py [hunk 7]
```diff
@@ -168,7 +200,7 @@ def _update_settings(self, X):
         :type X: list of (candidate, features) pairs
         """
 
-        self.logger.info("Load defalut parameters for Sparse LSTM")
+        self.logger.info("Loading defalut parameters for Sparse LSTM")
```
Comment (lukehsiao) [inline src/fonduer/learning/disc_models/sparse_lstm.py]:
```suggestion
        self.logger.info("Loading default parameters for Sparse LSTM")
```

File: src/fonduer/learning/disc_models/logistic_regression.py [hunk 8]
```diff
@@ -78,7 +94,7 @@ def _update_settings(self, X):
         :type X: list of (candidate, features) pair
         """
 
-        self.logger.info("Load defalut parameters for Logistic Regression")
+        self.logger.info("Loading defalut parameters for Logistic Regression")
```
Comment (lukehsiao) [inline src/fonduer/learning/disc_models/logistic_regression.py]:
```suggestion
        self.logger.info("Loading default parameters for Logistic Regression")
```
